src/ctr_logging.c

```diff
@@ -246,30 +256,47 @@ int write_journald(int pipe, char *buf, ssize_t buflen)
 	 * to be changed if this convention is changed.
 	 */
 	const char *message_priority = "PRIORITY=6";
-	if (pipe == STDERR_PIPE)
+	if (pipe == STDERR_PIPE) {
 		message_priority = "PRIORITY=3";
+		partial_buf = stderr_partial_buf;
+		partial_buf_len = &stderr_partial_buf_len;
+	} else {
+		partial_buf = stdout_partial_buf;
+		partial_buf_len = &stdout_partial_buf_len;
+	}
 
 	ptrdiff_t line_len = 0;
 
-	while (buflen > 0) {
+	while (buflen > 0 || *partial_buf_len > 0) {
 		writev_buffer_t bufv = {0};
 
-		bool partial = get_line_len(&line_len, buf, buflen);
+		bool partial = buflen == 0 || get_line_len(&line_len, buf, buflen);
+
+		/* If this is a partial line, and we have capacity to buffer it, buffer it and return.
+		 * The capacity of the partial_buf is one less than its size so that we can always add
+		 * a null terminating char later */
+		if (buflen && partial && ((unsigned long)line_len < (STDIO_BUF_SIZE - *partial_buf_len))) {
+			memcpy(partial_buf + *partial_buf_len, buf, line_len);
+			*partial_buf_len += line_len;
+			return 0;
+		}
+
 		/* sd_journal_* doesn't have an option to specify the number of bytes to write in the message, and instead writes the
 		 * entire string. Copying every line doesn't make very much sense, so instead we do this tmp_line_end
 		 * hack to emulate separate strings.
 		 */
 		char tmp_line_end = buf[line_len];
 		buf[line_len] = '\0';
 
-		_cleanup_free_ char *message = g_strdup_printf("MESSAGE=%s", buf);
-		if (writev_buffer_append_segment(dev_null, &bufv, message, line_len + MESSAGE_EQ_LEN) < 0)
+		ssize_t msg_len = line_len + MESSAGE_EQ_LEN + *partial_buf_len;
+		partial_buf[*partial_buf_len] = '\0';
+		_cleanup_free_ char *message = g_strdup_printf("MESSAGE=%s%s", partial_buf, buf);
+		if (writev_buffer_append_segment(dev_null, &bufv, message, msg_len) < 0)
```

Comment:
would it work if you don't allocate a string here (I know it is pre-existing code) but now we are already having an extra buffered copy:
```
if (writev_buffer_append_segment(dev_null, &bufv, "MESSAGE=", MESSAGE_EQ_LEN) < 0)
....
if (writev_buffer_append_segment(dev_null, &bufv, partial_buf, *partial_buf_len) < 0)
...
if (writev_buffer_append_segment(dev_null, &bufv, buf, line_len) < 0)
...
```

Reply:
Each structure in the iovec that is passed to sd_journal_sendv must represent one field of the entry to submit. What you have above would create 3 separate structures in the iovec for the MESSAGE field, so that doesn't quite work. Ultimately all components of the entry need to be stored in contiguous memory, one way or another. It would be possible to avoid the heap allocation by placing the message field in a char array on the stack (and memcpy the components of the message there). What do you think?